src/plugins/dashboard/public/types.ts

```diff
@@ -6,109 +6,184 @@
+ * update, and view its state.
  */
-export type DashboardAppStateInUrl = Omit<DashboardAppState, 'panels'> & {
-  panels?: SavedDashboardPanel[];
-};
-
-export interface DashboardAppStateTransitions {
-  set: (
-    state: DashboardAppState
-  ) => <T extends keyof DashboardAppState>(
-    prop: T,
-    value: DashboardAppState[T]
-  ) => DashboardAppState;
-  setOption: (
-    state: DashboardAppState
-  ) => <T extends keyof DashboardAppState['options']>(
-    prop: T,
-    value: DashboardAppState['options'][T]
-  ) => DashboardAppState;
+export interface DashboardAppState {
+  hasUnsavedChanges?: boolean;
+  indexPatterns?: IndexPattern[];
+  updateLastSavedState?: () => void;
+  resetToLastSavedState?: () => void;
+  savedDashboard?: DashboardSavedObject;
+  dashboardContainer?: DashboardContainer;
+  getLatestDashboardState?: () => DashboardState;
+  $triggerDashboardRefresh: Subject<{ force?: boolean }>;
+  $onDashboardStateChange: BehaviorSubject<DashboardState>;
+  applyFilters?: (query: Query, filters: Filter[]) => void;
 }
 
-export interface SavedDashboardPanelMap {
-  [key: string]: SavedDashboardPanel;
+/**
+ * The shared services and tools used to build a dashboard from a saved object ID.
+ */
+export interface DashboardBuildContext {
+  history: History;
+  kibanaVersion: string;
+  savedDashboardId?: string;
+  isEmbeddedExternally: boolean;
+  services: DashboardAppServices;
+  kbnUrlStateStorage: IKbnUrlStateStorage;
+  $checkForUnsavedChanges: Subject<unknown>;
+  getLatestDashboardState: () => DashboardState;
+  dispatchDashboardStateChange: Dispatch<AnyAction>;
+  $triggerDashboardRefresh: Subject<{ force?: boolean }>;
+  $onDashboardStateChange: BehaviorSubject<DashboardState>;
+}
+
+export interface DashboardOptions {
+  hidePanelTitles: boolean;
+  useMargins: boolean;
+  syncColors: boolean;
+}
+
+export type DashboardRedirect = (props: RedirectToProps) => void;
+export type RedirectToProps =
+  | { destination: 'dashboard'; id?: string; useReplace?: boolean; editMode?: boolean }
+  | { destination: 'listing'; filter?: string; useReplace?: boolean };
+
+export interface DashboardEmbedSettings {
+  forceHideFilterBar?: boolean;
+  forceShowTopNavMenu?: boolean;
+  forceShowQueryInput?: boolean;
+  forceShowDatePicker?: boolean;
+}
+
+export interface DashboardSaveOptions {
+  newTitle: string;
+  newTags?: string[];
+  newDescription: string;
+  newCopyOnSave: boolean;
+  newTimeRestore: boolean;
+  onTitleDuplicate: () => void;
+  isTitleDuplicateConfirmed: boolean;
+}
+
+export interface DashboardCapabilities {
+  show: boolean;
+  createNew: boolean;
+  saveQuery: boolean;
+  createShortUrl: boolean;
+  hideWriteControls: boolean;
+  storeSearchSession: boolean;
+  mapsCapabilities: { save: boolean };
+  visualizeCapabilities: { save: boolean };
 }
 
-export interface StagedFilter {
-  field: string;
-  value: string;
-  operator: string;
-  index: string;
+export interface DashboardAppServices {
```

Comment:
We should consider structuring this by concern, (though service abstractions may solve this entirely)... we've got event handlers mixed with components mixed with entire service API collections...?

Reply:
I totally agree about restructuring this. I am saving the refactor of everything services related for our next tech-debt removal session! Very excited to clean this up.